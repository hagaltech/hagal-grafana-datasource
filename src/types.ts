import {ChangeEvent} from 'react';
import {
  DataSourceJsonData,
  DataQueryRequest,
  QueryEditorProps,
  SelectableValue
} from '@grafana/data';

import {
  DataQuery,
} from '@grafana/schema';

import { DataSource } from './datasource';

export enum AggregatesType {
  AVERAGE = 'average',
  MAX = 'max',
  MIN = 'min',
  SUM = 'sum',
  COUNT = 'count',
  INTERPOLATION = 'interpolation',
}

// data source config options
export interface MyDataSourceOptions extends DataSourceJsonData {
  apiURL?: string;
  authDomain?: string;
}

// secure data source config options - values used in the backend, but never sent over HTTP to the frontend
export interface MySecureDataSourceOptions {
  clientId?: string;
  clientSecret?: string;
}

export interface DataSourceConnectionOptions extends MyDataSourceOptions, MySecureDataSourceOptions {
}

export const defaultMyDataSourceOptions = {
  apiURL: 'https://api.hagal.com/api/v1',
  authDomain: ''
};

// query interface options
export interface TimeSeriesQuery extends DataQuery {
  id: string;
  granularity?: string;
  label?: string;
  rawDataEnabled: boolean;
  name?: string;
  aggregation: AggregatesType;
}

export type OnQueryChange = (
  patch: Partial<TimeSeriesQuery>,
  shouldRunQuery?: boolean,
  withTimeout?: boolean,
) => void;

export type OnFieldChange = (
  key: keyof TimeSeriesQuery,
  shouldRunQuery?: boolean,
  withTimeout?: boolean,
) => (event: ChangeEvent<HTMLInputElement>) => void;

export type QueryOptions = DataQueryRequest<TimeSeriesQuery>;

export const defaultQueryOptions: Partial<TimeSeriesQuery> = {
  id: '',
  granularity: '2s',
  rawDataEnabled: false,
  label: '',
  name: '',
  aggregation: AggregatesType.AVERAGE
};

export type Tuple<T> = [T, T];

export enum HttpMethod {
  POST = 'POST',
  GET = 'GET',
}

export type EditorProps = QueryEditorProps<
  DataSource,
  TimeSeriesQuery,
  MyDataSourceOptions
>;

// proxy request options
export type DataQueryRequestItem = {
  start?: string | number;
  end?: string | number;
  limit?: number;
  id: string;
};

export type ProxyRequestDataOptions = {
  aggregates?: [AggregatesType];
  granularity?: string;
  ignoreUnknownIds?: boolean;
  items: DataQueryRequestItem[];
  limit?: number;
  start?: string | number;
  end?: string | number;
};

export const defaultProxyRequestDataOptions = {
  ignoreUnknownIds: false,
};

export type ProxyRequestOptions = {
  data?: ProxyRequestDataOptions;
  endpoint?: string;
  method: HttpMethod;
}

// proxy response
export interface ProxyResponseDataPointItem {
  [AggregatesType.AVERAGE]?: number;
  [AggregatesType.MAX]?: number;
  [AggregatesType.MIN]?: number;
  [AggregatesType.SUM]?: number;
  [AggregatesType.COUNT]?: number;
  [AggregatesType.INTERPOLATION]?: number;
  value?: number;
  timestamp?: number;
}

export interface ProxyResponseItem {
  id: string;
  name?: string;
}

export interface ProxyResponseDataItem extends ProxyResponseItem {
  datapoints: ProxyResponseDataPointItem[];
}

export interface ProxyResponseData {
  items: ProxyResponseDataItem[];
  id?: string;
}

export interface ProxyResponse {
  status: number,
  data: ProxyResponseData;
}

// time series data source types
export interface DataSourceRequestOptions extends ProxyRequestOptions {
  target: TimeSeriesQuery;
}

export type DataSourceItem = {
  data: ProxyResponseData;
  target: TimeSeriesQuery;
};

export type QueryProxyType = (options: ProxyRequestOptions) => Promise<ProxyResponse>;
export type GetSingleTimeseriesType = (id: string) => Promise<SelectableValue | null>;

export interface QueryRequestError {
  refId: string;
  requestId: string;
  error: string;
}

export type DropdownOptions = Promise<Array<SelectableValue<string>>>;

export type Range = {
  start: number;
  end: number;
  precomputed?: boolean;
};
