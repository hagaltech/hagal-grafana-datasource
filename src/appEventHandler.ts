import { EventBusSrv, BusEventWithPayload } from '@grafana/data';
import {stringifyError} from './utils';
import {QueryRequestError} from './types';

export const eventBusService = new EventBusSrv();

export const FAILED_ERROR_TYPE = 'failed-request';
export class FailedResponseEvent extends BusEventWithPayload<QueryRequestError> {
  static type = FAILED_ERROR_TYPE;
}
type EventClass<T> = new (payload: T) => BusEventWithPayload<T>;

function publishEvent<T>(EventClassType: EventClass<T>, payload: T) {
  const eventInstance = new EventClassType(payload);
  eventBusService.publish(eventInstance);
}

export function handleError(error: any, refId: string, requestId: string) {
  const errMessage = stringifyError(error);
  publishEvent(FailedResponseEvent, { refId, requestId, error: errMessage });
}
