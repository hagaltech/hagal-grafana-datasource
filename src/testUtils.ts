import {of, throwError} from 'rxjs';
import {DataSourceInstanceSettings} from '@grafana/data';
import _ from 'lodash';

import {DataSource, PROXY_PATH} from './datasource';
import {DataSourceConnectionOptions, ProxyRequestDataOptions} from './types';

import timeseriesFilterData from 'mock/timeseriesFilterData.json';

type MockData = {
  [key: string]: any;
};
type ErrorObject = {
  status?: number;
  data?: {
    error: {
      code: number;
      message: string;
    };
  };
};
type mockErrorFunctionType = ((data: ProxyRequestDataOptions) => ErrorObject) | undefined;

export function getItemsResponseObject(items: MockData) {
  return {
    data: {
      items,
    },
    status: 200
  };
}

export function getTimeseriesResponse(data: ProxyRequestDataOptions, dpNumber = 5) {
  const {aggregates, items} = data;
  const aggregate = aggregates ? aggregates[0] : 'value';

  const datapoints = new Array(dpNumber).fill(null).map((_, i) => ({
    timestamp: i * 600000 + 1549336675000,
    [aggregate]: i,
  }));

  return _.map(items, ({ id }) => ({
    id,
    datapoints,
  }));
}

const instanceSettings: DataSourceInstanceSettings<DataSourceConnectionOptions> = {
  id: 1,
  type: 'hagal-api-datasource',
  name: 'Hagal API Data',
  access: 'direct'
} as DataSourceInstanceSettings<DataSourceConnectionOptions>;

export const getMockedDataSource = () => new DataSource(instanceSettings);

const mockData: MockData = {
  [`/${PROXY_PATH}/timeseries/filter`]: () => timeseriesFilterData,
  [`/${PROXY_PATH}/datapoints/aggregate`]: getTimeseriesResponse,
  [`/${PROXY_PATH}/datapoints/list`]: getTimeseriesResponse,
};

export function getMockedFetch(mockErrorFunction?: mockErrorFunctionType) {
  return ({ url, data }: { url: string, data: ProxyRequestDataOptions }) => {
    if (mockErrorFunction) {
      const error = mockErrorFunction(data);
      if (error) {
        return throwError(() => error);
      }
    }

    if (mockData[url]) {
      return of(getItemsResponseObject(mockData[url](data)));
    }

    return of(getItemsResponseObject([]));
  };
}
