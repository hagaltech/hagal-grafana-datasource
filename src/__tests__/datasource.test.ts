import {DataQueryResponse, DataQueryResponseData} from '@grafana/data';

import {getMockedDataSource, getMockedFetch} from 'testUtils';
import {AggregatesType, TimeSeriesQuery} from '../types';
import {eventBusService, FAILED_ERROR_TYPE} from '../appEventHandler';

const mockedFetch = jest.fn(getMockedFetch());

// Spying on 'publish' method to track handleError calls from the appEventHandler.ts
const mockedErrorEventHandler = jest.spyOn(eventBusService, 'publish');
jest.mock('@grafana/runtime', () => ({
  getBackendSrv: jest.fn(() => ({
    fetch: mockedFetch
  }))
}));

const ds = getMockedDataSource();

const options: any = {
  targets: [],
  range: {
    from: 1549336675000,
    to: 1549338475000,
  },
  interval: '30s',
  intervalMs: 30000,
  maxDataPoints: 491,
  format: 'json',
  panelId: 1,
  dashboardId: 1,
};

const granularityError = {
  status: 400,
  data: {
    error: {
      code: 400,
      message: 'Unrecognized granularity',
    },
  },
};

type Mock = jest.Mock;
type TimeSeriesQueryLike = Partial<TimeSeriesQuery>;
function getFieldByName(data: DataQueryResponseData, name: string) {
  return data.fields.find((field: { name: string }) => field.name === name);
}
describe('Datasource Query', () => {
  describe('Given "Search Timeseries" queries', () => {
    let result: DataQueryResponse;
    const label = 'test-label';
    const id1 = 'test1';
    const id4 = 'test4';

    const tsTargetA: TimeSeriesQueryLike = {
      aggregation: AggregatesType.AVERAGE,
      refId: 'A',
      id: id1,
      granularity: '1h'
    };
    const tsTargetB: TimeSeriesQueryLike = {
      aggregation: AggregatesType.MIN,
      refId: 'B',
      id: 'test2',
      granularity: '1h',
    };
    const tsTargetC: TimeSeriesQueryLike = {
      aggregation: AggregatesType.MAX,
      refId: 'C',
      id: 'test3',
      granularity: '1h',
      label
    };
    const tsTargetD: TimeSeriesQueryLike = {
      refId: 'C',
      id: id4,
      granularity: '1h',
      rawDataEnabled: true
    };

    beforeAll(async () => {
      options.intervalMs = 1;
      options.targets = [tsTargetA, tsTargetB, tsTargetC, tsTargetD];

      result = await ds.query(options);
    });

    it('should generate the correct queries', async () => {
      // the first is for setOptionsForDropdown
      expect(mockedFetch).toBeCalledTimes(5);
      expect((mockedFetch as Mock).mock.calls[1][0]).toMatchSnapshot();
      expect((mockedFetch as Mock).mock.calls[2][0]).toMatchSnapshot();
      expect((mockedFetch as Mock).mock.calls[3][0]).toMatchSnapshot();
      expect((mockedFetch as Mock).mock.calls[4][0]).toMatchSnapshot();
    });

    it('should return the correct datapoints', () => {
      const { id: idA } = tsTargetA;
      const { id: idB } = tsTargetB;
      const { id: idC } = tsTargetC;
      const { id: idD } = tsTargetD;

      const targetFieldA = getFieldByName(result.data[0], 'target');
      const targetFieldB = getFieldByName(result.data[1], 'target');
      const targetFieldC = getFieldByName(result.data[2], 'target');
      const targetFieldD = getFieldByName(result.data[3], 'target');

      expect(targetFieldA).toBeDefined();
      targetFieldA && expect(targetFieldA.values.get(0)).toEqual(idA);

      expect(targetFieldB).toBeDefined();
      targetFieldB && expect(targetFieldB.values.get(0)).toEqual(idB);

      expect(targetFieldC).toBeDefined();
      targetFieldC && expect(targetFieldC.values.get(0)).toEqual(idC);

      expect(targetFieldD).toBeDefined();
      targetFieldD && expect(targetFieldD.values.get(0)).toEqual(idD);
    });

    it('should return correct labels', () => {
      const labelFieldC = getFieldByName(result.data[2], label);

      expect(labelFieldC).toBeDefined();
      expect(labelFieldC.values.length).toBeGreaterThan(1);
      expect(labelFieldC.values.every((value: number) => value !== undefined)).toBeTruthy();
    });

    it('should return correct datapoints with rawDataEnabled', () => {
      const labelFieldA = getFieldByName(result.data[0], id1);
      expect(labelFieldA).toBeDefined();
      expect(labelFieldA.values.length).toBeGreaterThan(1);
      expect(labelFieldA.values.every((value: number) => value !== undefined)).toBeTruthy();

      const labelFieldD = getFieldByName(result.data[3], id4);
      expect(labelFieldD).toBeDefined();
      expect(labelFieldD.values.length).toBeGreaterThan(1);
      expect(labelFieldD.values.every((value: number) => value !== undefined)).toBeTruthy();
    });
  });

  describe('Given "Search Timeseries" queries with errors', () => {
    let result: DataQueryResponse;
    const id = 'test1';
    const tsTargetA: TimeSeriesQueryLike = {
      aggregation: AggregatesType.AVERAGE,
      refId: 'A',
      id,
    };
    const tsTargetB: TimeSeriesQueryLike = {
      aggregation: AggregatesType.AVERAGE,
      refId: 'B',
      id: 'test2',
    };

    beforeAll(async () => {
      mockedFetch.mockImplementation(getMockedFetch(data => data.items[0].id === id ? granularityError : {}));

      options.intervalMs = 60000; // 1m
      options.targets = [tsTargetA, tsTargetB];

      result = await ds.query(options);
    });

    it('should return an empty array', () => {
      expect(result).toEqual({ data: [] });
    });

    it('should display errors for invalid queries', () => {
      expect(mockedErrorEventHandler as Mock).toBeCalledTimes(2);

      expect((mockedErrorEventHandler as Mock).mock.calls[0][0]).toEqual({
        type: FAILED_ERROR_TYPE,
        payload: {
          refId: 'A',
          error: '[400 ERROR] Unrecognized granularity',
        }
      });

      expect((mockedErrorEventHandler as Mock).mock.calls[1][0]).toEqual({
        type: FAILED_ERROR_TYPE,
        payload: {
          refId: 'B',
          error: 'Unknown error',
        }
      });
    });
  });
});
