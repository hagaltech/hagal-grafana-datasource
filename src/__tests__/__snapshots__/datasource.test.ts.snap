// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datasource Query Given "Search Timeseries" queries should generate the correct queries 1`] = `
{
  "data": {
    "aggregates": [
      "average",
    ],
    "end": 1549338475000,
    "granularity": "1h",
    "ignoreUnknownIds": false,
    "items": [
      {
        "id": "test1",
      },
    ],
    "limit": 3000,
    "start": 1549336675000,
  },
  "method": "POST",
  "url": "/api/datapoints/aggregate",
}
`;

exports[`Datasource Query Given "Search Timeseries" queries should generate the correct queries 2`] = `
{
  "data": {
    "aggregates": [
      "min",
    ],
    "end": 1549338475000,
    "granularity": "1h",
    "ignoreUnknownIds": false,
    "items": [
      {
        "id": "test2",
      },
    ],
    "limit": 3000,
    "start": 1549336675000,
  },
  "method": "POST",
  "url": "/api/datapoints/aggregate",
}
`;

exports[`Datasource Query Given "Search Timeseries" queries should generate the correct queries 3`] = `
{
  "data": {
    "aggregates": [
      "max",
    ],
    "end": 1549338475000,
    "granularity": "1h",
    "ignoreUnknownIds": false,
    "items": [
      {
        "id": "test3",
      },
    ],
    "limit": 3000,
    "start": 1549336675000,
  },
  "method": "POST",
  "url": "/api/datapoints/aggregate",
}
`;

exports[`Datasource Query Given "Search Timeseries" queries should generate the correct queries 4`] = `
{
  "data": {
    "ignoreUnknownIds": false,
    "items": [
      {
        "end": 1549338475000,
        "id": "test4",
        "limit": 0,
        "start": 1549336675000,
      },
    ],
  },
  "method": "POST",
  "url": "/api/datapoints/list",
}
`;
